src/Symfony/Component/Security/Csrf/DoubleSubmitCsrfTokenManager.php

```diff
@@ -0,0 +1,247 @@
+
+        foreach ($request->cookies->all() as $name => $value) {
+            if ($this->cookieName === $value && str_starts_with($name, $cookieName.'_')) {
+                $response->headers->clearCookie($name, '/', null, $request->isSecure(), false, 'strict');
+            }
+        }
+    }
+
+    public function persistStrategy(Request $request): void
+    {
+        if ($request->hasSession(true) && $request->attributes->has($this->cookieName)) {
+            $request->getSession()->set($this->cookieName, $request->attributes->get($this->cookieName));
+        }
+    }
+
+    public function onKernelResponse(ResponseEvent $event): void
+    {
+        if (!$event->isMainRequest()) {
+            return;
+        }
+
+        $this->clearCookie($event->getRequest(), $event->getResponse());
+        $this->persistStrategy($event->getRequest());
+    }
+
+    /**
+     * @return bool|null Whether the origin is valid, null if missing
+     */
+    private function isValidOrigin(Request $request): ?bool
+    {
+        $source = $request->headers->get('Origin') ?? $request->headers->get('Referer') ?? 'null';
+
+        return 'null' === $source ? null : str_starts_with($source.'/', $request->getSchemeAndHttpHost().'/');
+    }
+
+    /**
+     * @return bool|null Whether the double-submit is valid, null if missing
+     */
+    private function isValidDoubleSubmit(Request $request, string $token): ?bool
+    {
+        $cookieName = ($request->isSecure() ? '__Host-' : '').$this->cookieName;
+
+        if (self::CHECK_ONLY_HEADER === $this->checkHeader) {
+            if ($request->headers->get($this->cookieName) === $token) {
+                $request->cookies->set($cookieName.'_'.$token, $this->cookieName); // Ensure clearCookie() can remove any cookie filtered by a reverse-proxy
+                $this->logger?->debug('CSRF validation accepted using header info.');
+
+                return true;
+            }
+
+            if ($request->headers->has($this->cookieName)) {
+                $this->logger?->warning('CSRF validation failed: wrong token found in header info.');
+
+                return false;
+            }
+
+            return null;
+        }
+
+        if (!isset($request->cookies->all()[$cookieName.'_'.$token]) && !($this->checkHeader && $request->headers->has($this->cookieName))) {
+            return null;
+        }
+
+        if ($this->checkHeader && $request->headers->get($this->cookieName) !== $token) {
+            $this->logger?->warning('CSRF validation failed: wrong token found in header info.');
+
+            return false;
+        }
+
+        if ($request->cookies->all()[$cookieName.'_'.$token] !== $this->cookieName) {
+            $this->logger?->warning('CSRF validation failed: token not found in cookies.');
+
+            return false;
+        }
+
+        $this->logger?->debug('CSRF validation accepted using double-submit info.');
+
+        return true;
```

Comment:
I don't like this default option. the user is sensible to an attack until the cookie is cleared.

attack vector:
 - if a legitimate request fails -> segfault or 503 or timeout or whatever -> the cookie stays and the door is open for an attacker
 - legitimate request is slow -> door open until the response arrives to the client
 - attacker run request in a loop, and wait for the victim to perform a legitimate request that put the cookie demand open the door.

Reply:
This is not a fallback: if we reach this point, the double-submit is verified.

I don't think I see the concern you're raising here, can you expand?

Even if I'm unsure about it, your comment made me think about two changes:
- when turbo is used, we clear the cookie in the browser thanks to the `turbo:submit-end` event, see https://github.com/symfony/recipes/pull/1337
- we now store in the session whether both origin+double-submit checks succeeded, or if just one of them did, and we fail if next requests don't provide at least the same level.

Reply:
@jderusse because the cookie is unique and randomly generated, I don't see which door is open during e.g. slow requests.